bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/UncleanShutdownDetectionImpl.java

```diff
@@ -0,0 +1,87 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.bookkeeper.bookie;
+
+import java.io.File;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Used to determine if the prior shutdown was unclean or not. It does so
+ * by adding a file to each ledger directory after successful start-up
+ * and removing the file on graceful shutdown.
+ * Any abrupt termination will cause one or more of these files to not be cleared
+ * and so on the subsequent boot-up, the presence of any of these files will
+ * indicate an unclean shutdown.
+ */
+public class UncleanShutdownDetectionImpl implements UncleanShutdownDetection {
+    private static final Logger LOG = LoggerFactory.getLogger(UncleanShutdownDetectionImpl.class);
+    private final LedgerDirsManager ledgerDirsManager;
+    static final String DirtyFileName = "DIRTY";
+
+    public UncleanShutdownDetectionImpl(LedgerDirsManager ledgerDirsManager) {
+        this.ledgerDirsManager = ledgerDirsManager;
+    }
+
+    @Override
+    public void registerStartUp() {
+        for (File ledgerDir : ledgerDirsManager.getAllLedgerDirs()) {
+            try {
+                File dirtyFile = new File(ledgerDir, DirtyFileName);
+                dirtyFile.createNewFile();
+            } catch (Throwable t) {
+                LOG.error("Unable to register start-up (so an unclean shutdown cannot"
+                        + " be detected). Dirty file of ledger dir {} could not be created",
+                        ledgerDir.getAbsolutePath(), t);
+            }
+        }
+    }
+
+    @Override
+    public void registerCleanShutdown() {
+        for (File ledgerDir : ledgerDirsManager.getAllLedgerDirs()) {
+            try {
+                File dirtyFile = new File(ledgerDir, DirtyFileName);
+                dirtyFile.delete();
+            } catch (Throwable t) {
+                LOG.error("Unable to register a clean shutdown, dirty file of "
+                        + " ledger dir {} could not be deleted",
+                        ledgerDir.getAbsolutePath(), t);
+            }
+        }
+    }
+
+    @Override
+    public boolean lastShutdownWasUnclean() {
+        try {
+            for (File ledgerDir : ledgerDirsManager.getAllLedgerDirs()) {
+                File dirtyFile = new File(ledgerDir, DirtyFileName);
+                if (dirtyFile.exists()) {
+                    return true;
```

Comment:
Well, we log the detection of an unclean shutdown in the Main class, but I see that it could be useful to know which ledger directory did not get shutdown cleanly. But if we want to go that route, we'd want to collect all the directories with a dirty file and log them, rather returning on the first.

Reply:
I've updated it to do that.